File: src/typescript/src/shared/llm-client.ts
/**
 * Shared LLM Client for TypeScript Orchestration Patterns
 *
 * Provides a unified interface for calling Claude with:
 * - Token tracking
 * - Latency measurement
 * - Cost calculation
 */

import Anthropic from "@anthropic-ai/sdk";
import { config } from "dotenv";
import { resolve } from "path";

// Load environment variables from config/.env
config({ path: resolve(process.cwd(), "../../config/.env") });

// Also try the project root
config({ path: resolve(process.cwd(), "../../../config/.env") });

export interface LLMResponse {
  content: string;
  inputTokens: number;
  outputTokens: number;
  latencyMs: number;
  model: string;
}

export interface AgentConfig {
  name: string;
  systemPrompt: string;
  temperature?: number;
  maxTokens?: number;
}

export class LLMClient {
  private client: Anthropic;
  private model: string = "claude-3-haiku-20240307";

  // Pricing per 1M tokens (as of 2024)
  private pricing = {
    "claude-3-haiku-20240307": { input: 0.25, output: 1.25 },
    "claude-3-sonnet-20240229": { input: 3.0, output: 15.0 },
    "claude-3-opus-20240229": { input: 15.0, output: 75.0 },
  };

  constructor(model?: string) {
    const apiKey = process.env.ANTHROPIC_API_KEY;
    if (!apiKey) {
      throw new Error(
        "ANTHROPIC_API_KEY not found. Set it in config/.env or environment."
      );
    }

    this.client = new Anthropic({ apiKey });
    if (model) {
      this.model = model;
    }
  }

  async complete(
    prompt: string,
    options: {
      systemPrompt?: string;
      temperature?: number;
      maxTokens?: number;
    } = {}
  ): Promise<LLMResponse> {
    const startTime = performance.now();

    const response = await this.client.messages.create({
      model: this.model,
      max_tokens: options.maxTokens ?? 1024,
      temperature: options.temperature ?? 0.7,
      system: options.systemPrompt,
      messages: [{ role: "user", content: prompt }],
    });

    const latencyMs = performance.now() - startTime;

    const content =
      response.content[0].type === "text" ? response.content[0].text : "";

    return {
      content,
      inputTokens: response.usage.input_tokens,
      outputTokens: response.usage.output_tokens,
      latencyMs,
      model: this.model,
    };
  }

  async chat(
    messages: Array<{ role: "user" | "assistant"; content: string }>,
    options: {
      systemPrompt?: string;
      temperature?: number;
      maxTokens?: number;
    } = {}
  ): Promise<LLMResponse> {
    const startTime = performance.now();

    const response = await this.client.messages.create({
      model: this.model,
      max_tokens: options.maxTokens ?? 1024,
      temperature: options.temperature ?? 0.7,
      system: options.systemPrompt,
      messages,
    });

    const latencyMs = performance.now() - startTime;

    const content =
      response.content[0].type === "text" ? response.content[0].text : "";

    return {
      content,
      inputTokens: response.usage.input_tokens,
      outputTokens: response.usage.output_tokens,
      latencyMs,
      model: this.model,
    };
  }

  calculateCost(inputTokens: number, outputTokens: number): number {
    const modelPricing =
      this.pricing[this.model as keyof typeof this.pricing] ||
      this.pricing["claude-3-haiku-20240307"];

    return (
      (inputTokens / 1_000_000) * modelPricing.input +
      (outputTokens / 1_000_000) * modelPricing.output
    );
  }

  getModel(): string {
    return this.model;
  }
}

/**
 * Agent class for multi-agent orchestration
 */
export class Agent {
  private client: LLMClient;
  private config: AgentConfig;
  private history: Array<{ role: "user" | "assistant"; content: string }> = [];

  // Track cumulative stats
  public totalInputTokens = 0;
  public totalOutputTokens = 0;
  public totalLatencyMs = 0;
  public callCount = 0;

  constructor(client: LLMClient, config: AgentConfig) {
    this.client = client;
    this.config = config;
  }

  get name(): string {
    return this.config.name;
  }

  async run(input: string): Promise<string> {
    const response = await this.client.complete(input, {
      systemPrompt: this.config.systemPrompt,
      temperature: this.config.temperature ?? 0.7,
      maxTokens: this.config.maxTokens ?? 1024,
    });

    // Track stats
    this.totalInputTokens += response.inputTokens;
    this.totalOutputTokens += response.outputTokens;
    this.totalLatencyMs += response.latencyMs;
    this.callCount++;

    return response.content;
  }

  async chat(input: string): Promise<string> {
    // Add user message to history
    this.history.push({ role: "user", content: input });

    const response = await this.client.chat(this.history, {
      systemPrompt: this.config.systemPrompt,
      temperature: this.config.temperature ?? 0.7,
      maxTokens: this.config.maxTokens ?? 1024,
    });

    // Add assistant response to history
    this.history.push({ role: "assistant", content: response.content });

    // Track stats
    this.totalInputTokens += response.inputTokens;
    this.totalOutputTokens += response.outputTokens;
    this.totalLatencyMs += response.latencyMs;
    this.callCount++;

    return response.content;
  }

  clearHistory(): void {
    this.history = [];
  }

  getStats(): {
    name: string;
    calls: number;
    inputTokens: number;
    outputTokens: number;
    latencyMs: number;
  } {
    return {
      name: this.config.name,
      calls: this.callCount,
      inputTokens: this.totalInputTokens,
      outputTokens: this.totalOutputTokens,
      latencyMs: this.totalLatencyMs,
    };
  }
}

/**
 * Metrics tracker for orchestration patterns
 */
export class OrchestrationMetrics {
  private agents: Agent[] = [];
  private startTime: number = 0;
  private endTime: number = 0;

  addAgent(agent: Agent): void {
    this.agents.push(agent);
  }

  start(): void {
    this.startTime = performance.now();
  }

  end(): void {
    this.endTime = performance.now();
  }

  getSummary(): {
    totalLatencyMs: number;
    totalInputTokens: number;
    totalOutputTokens: number;
    totalCalls: number;
    agentStats: ReturnType<Agent["getStats"]>[];
  } {
    const agentStats = this.agents.map((a) => a.getStats());

    return {
      totalLatencyMs: this.endTime - this.startTime,
      totalInputTokens: agentStats.reduce((sum, a) => sum + a.inputTokens, 0),
      totalOutputTokens: agentStats.reduce((sum, a) => sum + a.outputTokens, 0),
      totalCalls: agentStats.reduce((sum, a) => sum + a.calls, 0),
      agentStats,
    };
  }

  printSummary(patternName: string): void {
    const summary = this.getSummary();

    console.log("\n" + "=".repeat(70));
    console.log(`ORCHESTRATION METRICS: ${patternName}`);
    console.log("=".repeat(70));
    console.log(`\nTotal Latency: ${(summary.totalLatencyMs / 1000).toFixed(2)}s`);
    console.log(`Total LLM Calls: ${summary.totalCalls}`);
    console.log(`Total Input Tokens: ${summary.totalInputTokens}`);
    console.log(`Total Output Tokens: ${summary.totalOutputTokens}`);

    console.log("\nPer-Agent Breakdown:");
    console.log("-".repeat(60));
    for (const agent of summary.agentStats) {
      console.log(
        `  ${agent.name}: ${agent.calls} calls, ` +
          `${agent.inputTokens}/${agent.outputTokens} tokens, ` +
          `${(agent.latencyMs / 1000).toFixed(2)}s`
      );
    }
  }
}

File: src/python/shared/llm_client.py
"""
Unified LLM Client for OpenAI and Anthropic APIs.

This module provides a consistent interface for calling different LLM providers,
making it easy to compare responses and switch between models.
"""

import os
import time
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Optional, List, Dict, Any

from dotenv import load_dotenv
from pathlib import Path

# Load environment variables from config/.env
_project_root = Path(__file__).parent.parent.parent.parent
_env_path = _project_root / "config" / ".env"
load_dotenv(_env_path)


@dataclass
class LLMResponse:
    """Standardized response from any LLM provider."""
    content: str
    model: str
    provider: str
    input_tokens: int
    output_tokens: int
    total_tokens: int
    latency_ms: float
    cost_usd: float
    raw_response: Any = None


class LLMClient(ABC):
    """Abstract base class for LLM clients."""

    @abstractmethod
    def complete(
        self,
        prompt: str,
        system: Optional[str] = None,
        temperature: float = 0.0,
        max_tokens: int = 1024,
    ) -> LLMResponse:
        """Generate a completion for the given prompt."""
        pass

    @abstractmethod
    def get_name(self) -> str:
        """Return the provider/model name."""
        pass


class OpenAIClient(LLMClient):
    """OpenAI API client wrapper."""

    # Pricing per 1M tokens (as of 2024)
    PRICING = {
        "gpt-4o": {"input": 2.50, "output": 10.00},
        "gpt-4o-mini": {"input": 0.15, "output": 0.60},
        "gpt-4-turbo": {"input": 10.00, "output": 30.00},
        "gpt-3.5-turbo": {"input": 0.50, "output": 1.50},
    }

    def __init__(self, model: str = None, api_key: str = None):
        from openai import OpenAI

        self.model = model or os.getenv("OPENAI_MODEL", "gpt-4o-mini")
        self.client = OpenAI(api_key=api_key or os.getenv("OPENAI_API_KEY"))

    def complete(
        self,
        prompt: str,
        system: Optional[str] = None,
        temperature: float = 0.0,
        max_tokens: int = 1024,
    ) -> LLMResponse:
        messages = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})

        start_time = time.perf_counter()
        response = self.client.chat.completions.create(
            model=self.model,
            messages=messages,
            temperature=temperature,
            max_tokens=max_tokens,
        )
        latency_ms = (time.perf_counter() - start_time) * 1000

        # Calculate cost
        pricing = self.PRICING.get(self.model, {"input": 0, "output": 0})
        input_tokens = response.usage.prompt_tokens
        output_tokens = response.usage.completion_tokens
        cost = (input_tokens * pricing["input"] + output_tokens * pricing["output"]) / 1_000_000

        return LLMResponse(
            content=response.choices[0].message.content,
            model=self.model,
            provider="openai",
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            total_tokens=response.usage.total_tokens,
            latency_ms=latency_ms,
            cost_usd=cost,
            raw_response=response,
        )

    def get_name(self) -> str:
        return f"OpenAI/{self.model}"


class AnthropicClient(LLMClient):
    """Anthropic API client wrapper."""

    # Pricing per 1M tokens (as of 2024)
    PRICING = {
        "claude-sonnet-4-20250514": {"input": 3.00, "output": 15.00},
        "claude-3-5-sonnet-20241022": {"input": 3.00, "output": 15.00},
        "claude-3-haiku-20240307": {"input": 0.25, "output": 1.25},
        "claude-3-opus-20240229": {"input": 15.00, "output": 75.00},
    }

    def __init__(self, model: str = None, api_key: str = None):
        from anthropic import Anthropic

        self.model = model or os.getenv("ANTHROPIC_MODEL", "claude-3-haiku-20240307")
        self.client = Anthropic(api_key=api_key or os.getenv("ANTHROPIC_API_KEY"))

    def complete(
        self,
        prompt: str,
        system: Optional[str] = None,
        temperature: float = 0.0,
        max_tokens: int = 1024,
    ) -> LLMResponse:
        kwargs = {
            "model": self.model,
            "max_tokens": max_tokens,
            "temperature": temperature,
            "messages": [{"role": "user", "content": prompt}],
        }
        if system:
            kwargs["system"] = system

        start_time = time.perf_counter()
        response = self.client.messages.create(**kwargs)
        latency_ms = (time.perf_counter() - start_time) * 1000

        # Calculate cost
        pricing = self.PRICING.get(self.model, {"input": 0, "output": 0})
        input_tokens = response.usage.input_tokens
        output_tokens = response.usage.output_tokens
        cost = (input_tokens * pricing["input"] + output_tokens * pricing["output"]) / 1_000_000

        return LLMResponse(
            content=response.content[0].text,
            model=self.model,
            provider="anthropic",
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            total_tokens=input_tokens + output_tokens,
            latency_ms=latency_ms,
            cost_usd=cost,
            raw_response=response,
        )

    def get_name(self) -> str:
        return f"Anthropic/{self.model}"


def get_client(provider: str = "openai", model: str = None) -> LLMClient:
    """Factory function to get an LLM client by provider name."""
    if provider.lower() == "openai":
        return OpenAIClient(model=model)
    elif provider.lower() in ("anthropic", "claude"):
        return AnthropicClient(model=model)
    else:
        raise ValueError(f"Unknown provider: {provider}")


# Quick test
if __name__ == "__main__":
    print("Testing LLM clients...")

    # Test OpenAI
    try:
        openai_client = OpenAIClient()
        response = openai_client.complete("Say 'Hello, World!' and nothing else.")
        print(f"\nOpenAI Response: {response.content}")
        print(f"  Tokens: {response.total_tokens}, Cost: ${response.cost_usd:.6f}, Latency: {response.latency_ms:.0f}ms")
    except Exception as e:
        print(f"OpenAI error: {e}")

    # Test Anthropic
    try:
        anthropic_client = AnthropicClient()
        response = anthropic_client.complete("Say 'Hello, World!' and nothing else.")
        print(f"\nAnthropic Response: {response.content}")
        print(f"  Tokens: {response.total_tokens}, Cost: ${response.cost_usd:.6f}, Latency: {response.latency_ms:.0f}ms")
    except Exception as e:
        print(f"Anthropic error: {e}")
